issue/200 API v2 Maven Module

hhund opened the pull request:
New api infrastructure, api v2 module and upgrade to HAPI 7.4.0.

* new process plugin api v2 module with dependency to HAPI FHIR 7.4.0, old process plugin api v1 module still has dependency to HAPI FHIR 5.1.0
* HAPI dependency versions now managed via properties hapi.fhir.version and hapi.fhir.version.v1
* upgraded HAPI FHIR dependency of DSF FHIR and DSF BPE application to 7.4.0, remove not needed workarounds and added a new ones to SnapshotGeneratorImpl and ResourceValidatorImpl in dsf-fhir-validation
* added Task profile test to dsf-fhir-validation
* added missing thymeleaf dependency to dsf-common-ui module
* refactored code to use switch expressions with pattern matching where possible
* set log level for ca.uhn.fhir.parser.LenientErrorHandler to ERROR to suppress "Unknown element 'author' found while parsing" messages generated during startup
* removed dsf dependencies from process plugin api v1 module and copied dependend code into the module
* to allow v1 and v2 to have different HAPI versions as dependencies, no code with dependencies to classes from HAPI can be loaded by the BPE app class loader as theses classes could not be linked to different HAPI versions via the v1 or v2 api class loaders. Unfortunately this requirement leads to duplicated code between the api v1, api v2 and the bpe modules.

Known Issues / TODOs:
* The current ([v1.0.1.0](https://github.com/datasharingframework/dsf-process-ping-pong/releases/tag/v1.0.1.0)) Ping/Pong Process does not work with this branch. The reason for this is that the [Ping Status Extension in Line 37](
https://github.com/datasharingframework/dsf-process-ping-pong/blob/6312e7d8d0b7b2d5ab3eb2c8cbacd22ea70324c4/src/main/resources/fhir/StructureDefinition/dsf-extension-ping-status.xml#L37) has a bad `min = 3` configuration, now interpreted differently by the FHIR Validator of HAPI 7.4.0 resulting in invalid Task resources. We could implement a workaround that fixes the extension StructureDefinition or release a fixed process plugin.

closes #200
closes #201

hhund committed:
new api infrastructure and api v2 module

* new process plugin api v2 module with dependency to HAPI FHIR 7.4.0,
old process plugin api v1 module still has dependency to HAPI FHIR 5.1.0
* HAPI dependency versions now managed via properties hapi.fhir.version
and hapi.fhir.version.v1
* upgraded HAPI FHIR dependency of DSF FHIR and DSF BPE application to
7.4.0, remove not needed workarounds and added a new ones to
SnapshotGeneratorImpl and ResourceValidatorImpl in dsf-fhir-validation
* added Task profile test to dsf-fhir-validation
* added missing thymeleaf dependency to dsf-common-ui module
* refactored code to use switch expressions with pattern matching where
possible
* set log level for ca.uhn.fhir.parser.LenientErrorHandler to ERROR to
suppress "Unknown element 'author' found while parsing" messages
generated during startup

hhund committed:
code cleanup

hhund committed:
fix for #224

wetret commented on dsf-bpe/dsf-bpe-process-api-v2/src/main/java/dev/dsf/bpe/v2/activity/DefaultUserTaskListener.java:
```diff
@@ -0,0 +1,264 @@
+
+			updateFailedIfInprogress(variables.getTasks(), errorMessage);
+
+			// TODO evaluate throwing exception as alternative to stopping the process instance
```
Maybe create an issue for this todo. It could be resolved in api v2 release.

hhund replied:
Issue created, see #240

wetret commented on dsf-bpe/dsf-bpe-server-jetty/pom.xml:
```diff
@@ -89,6 +89,436 @@
+								<artifactItem>
+									<groupId>ca.uhn.hapi.fhir</groupId>
+									<artifactId>hapi-fhir-base</artifactId>
+									<version>7.4.0</version>
```
I assume that these are not properties, as the repository version numbers will not change for an API version.

hhund replied:
fixe via 3f539d0

wetret commented on dsf-bpe/dsf-bpe-server/src/main/java/dev/dsf/bpe/plugin/ProcessPluginApiClassLoader.java:
```diff
@@ -0,0 +1,193 @@
+	protected Class<?> loadClass(String className, boolean resolve) throws ClassNotFoundException
+	{
+		// TODO remove
+		// logger.trace("loadClass({}, {}) ...", className, resolve);
```
Remove, applies as well for TODOs further below

hhund replied:
Code reworked, issue for remaining work regarding class and resource allow-lists created, see #241

wetret commented on dsf-bpe/dsf-bpe-server/src/main/java/dev/dsf/bpe/plugin/ProcessPluginApiClassLoader.java:
```diff
@@ -0,0 +1,193 @@
+			return false;
+
+		logger.trace("TODO should class be hidden? {}", clazz.getName());
+		// TODO Auto-generated method stub
```
Remove, applies as well for TODOs further below

hhund replied:
Code reworked, issue for remaining work regarding class and resource allow-lists created, see #241

wetret commented on dsf-fhir/dsf-fhir-rest-adapter/src/main/java/dev/dsf/fhir/service/ReferenceExtractorImpl.java:
```diff
@@ -327,83 +327,57 @@ else if (streams.length == 2)
+			case ActivityDefinition ad -> getReferences(ad);
+
+			// not implemented yet, special rules apply for tmp ids
+			// case Bundle b -> getReferences(b);
```
maybe mark with TODO

hhund replied:
I removed the comment, an implementation for `Bundle` resources is not needed in the way we use the `ReferenceExtractor`.

wetret reviewed (approved):
Many thanks for the valuable adjustments and additions. Here are some comments.

hhund committed:
removed not needed TODO comment

HAPI bug workaround will never be removed from API v1 implementation.

hhund committed:
fixed bean method name

hhund committed:
removed not needed JsonAlias annotations

The JsonAlias annotations were introduced to allow deserialization of
older TargetImpl objects in DSF 0.x that used the alias values as
property names.

hhund commented on dsf-bpe/dsf-bpe-process-api-v1-impl/src/main/java/dev/dsf/bpe/v1/spring/ApiServiceConfig.java:
```diff
@@ -0,0 +1,200 @@
+	@Bean
+	public FhirContext fhirContext()
+	{
+		// TODO remove workaround after upgrading to HAPI 6.8+, see https://github.com/hapifhir/hapi-fhir/issues/5205
```
fixed via 704225a

hhund commented on dsf-bpe/dsf-bpe-process-api-v2-impl/src/main/java/dev/dsf/bpe/v2/spring/ApiServiceConfig.java:
```diff
@@ -0,0 +1,194 @@
+	private BpeMailService bpeMailService;
+
+	@Bean
+	public ProcessPluginApi processPluginApiV1()
```
fixed via c4eb826

hhund commented on dsf-bpe/dsf-bpe-process-api-v2-impl/src/main/java/dev/dsf/bpe/v2/variables/TargetImpl.java:
```diff
@@ -0,0 +1,61 @@
+	@JsonCreator
+	public TargetImpl(
+			@JsonProperty("organizationIdentifierValue") @JsonAlias("targetOrganizationIdentifierValue") String organizationIdentifierValue,
+			@JsonProperty("endpointIdentifierValue") String endpointIdentifierValue,
```
aliases no longer needed, longer explanation see commit message, removed via f95cae6

hhund committed:
hapi version property for api v2, version properties in bpe jetty pom

Introduced new HAPI version property for plugin API v2.
Replaced hard-coded HAPI version numbers in dsf-bpe-server-jetty and api
v2 modules with version properties.

hhund committed:
improved private method name

hhund committed:
simplified api class loader and added class/resource allow mechanism

* New ProcessPluginApiClassLoaderFactory to create
ProcessPluginApiClassLoader instances with allow lists for classes and
resources.
* Default allowed classes and resources stored in jar internal text
files for v1 and v2. Lists are incomplete for now

hhund committed:
removed TODO comment that will not be resolved for v1

hhund committed:
comment removed, implementation for Bundle resources not needed
